packages/loader/src/MaterialLoader.ts

```diff
@@ -32,73 +33,91 @@ class MaterialLoader extends Loader<Material> {
       })
         .then((materialSchema: IMaterialSchema) => {
           const engine = resourceManager.engine;
-          const { name, shader, shaderData, macros, renderState } = materialSchema;
-          const material = new Material(engine, Shader.find(shader));
-          material.name = name;
+          const { shaderRef, shader } = materialSchema;
 
-          const texturePromises = new Array<Promise<Texture2D>>();
-          const materialShaderData = material.shaderData;
-          for (let key in shaderData) {
-            const { type, value } = shaderData[key];
-
-            switch (type) {
-              case "Vector2":
-                materialShaderData.setVector2(key, new Vector2((<IVector2>value).x, (<IVector2>value).y));
-                break;
-              case "Vector3":
-                materialShaderData.setVector3(
-                  key,
-                  new Vector3((<IVector3>value).x, (<IVector3>value).y, (<IVector3>value).z)
-                );
-                break;
-              case "Vector4":
-                materialShaderData.setVector4(
-                  key,
-                  new Vector4((<IVector4>value).x, (<IVector4>value).y, (<IVector4>value).z, (<IVector4>value).w)
-                );
-                break;
-              case "Color":
-                materialShaderData.setColor(
-                  key,
-                  new Color((<IColor>value).r, (<IColor>value).g, (<IColor>value).b, (<IColor>value).a)
-                );
-                break;
-              case "Float":
-                materialShaderData.setFloat(key, <number>value);
-                break;
-              case "Texture":
-                texturePromises.push(
-                  // @ts-ignore
-                  resourceManager.getResourceByRef<Texture2D>(<IAssetRef>value).then((texture) => {
-                    materialShaderData.setTexture(key, texture);
-                  })
-                );
-                break;
-              case "Boolean":
-                materialShaderData.setInt(key, value ? 1 : 0);
-                break;
-              case "Integer":
-                materialShaderData.setInt(key, Number(value));
-                break;
-            }
+          if (shaderRef) {
+            resolve(
+              resourceManager
+                // @ts-ignore
+                .getResourceByRef<Shader>(<IAssetRef>shaderRef)
+                .then((shaderObject) => this.getMaterialByShader(materialSchema, shaderObject, engine))
+            );
+          } else {
```

Comment:
The builtin shader logic should go here, and should not be repeated in shaderLoader

Reply:
Shaders written with `ShaderLab` can still reference any of builtin shader

Reply:
* delete 'shaderRef' in editor if is builtinShader
* combine `#include` logic in shader、shaderChunk loader